src/messages.cpp

```diff
@@ -265,170 +265,586 @@ Sender::sender_type() const
+  : wire_format(wire_format_in)
+  , content(std::move(content_in))
+  , auth(std::move(auth_in))
+{}
+
+bytes
+MLSMessageContentAuth::to_be_signed(
+  const std::optional<GroupContext>& context) const
+{
+  return tls::marshal(MLSMessageContentTBS{
+    wire_format,
+    content,
+    context,
+  });
+}
+
+MLSPlaintext
+MLSPlaintext::protect(MLSMessageContentAuth content_auth,
+                      CipherSuite suite,
+                      const std::optional<bytes>& membership_key,
+                      const std::optional<GroupContext>& context)
 {
-  if (wire_format == WireFormat::mls_ciphertext) {
-    return true;
+  auto pt = MLSPlaintext(std::move(content_auth));
+
+  // Add the membership_mac if required
+  switch (pt.content.sender.sender_type()) {
+    case SenderType::member:
+      pt.membership_tag =
+        pt.membership_mac(suite, opt::get(membership_key), context);
+      break;
+
+    default:
+      break;
   }
 
-  if (!membership_tag) {
-    return false;
+  return pt;
+}
+
+std::optional<MLSMessageContentAuth>
+MLSPlaintext::unprotect(CipherSuite suite,
+                        const std::optional<bytes>& membership_key,
+                        const std::optional<GroupContext>& context) const
+{
+  // Verify the membership_tag if the message was sent within the group
+  switch (content.sender.sender_type()) {
+    case SenderType::member: {
+      auto candidate = membership_mac(suite, opt::get(membership_key), context);
+      if (candidate != opt::get(membership_tag)) {
+        return {};
+      }
+      break;
+    }
+
+    default:
+      break;
   }
 
-  return tag == opt::get(membership_tag);
+  return MLSMessageContentAuth{
+    WireFormat::mls_plaintext,
+    content,
+    auth,
+  };
+}
+
+MLSPlaintext::MLSPlaintext(MLSMessageContentAuth content_auth)
+  : content(std::move(content_auth.content))
+  , auth(std::move(content_auth.auth))
+{
+  if (content_auth.wire_format != WireFormat::mls_plaintext) {
+    throw InvalidParameterError("Wire format mismatch (not mls_plaintext)");
+  }
+}
+
+bytes
+MLSPlaintext::membership_mac(CipherSuite suite,
+                             const bytes& membership_key,
+                             const std::optional<GroupContext>& context) const
+{
+  auto tbm = tls::marshal(MLSMessageContentTBM{
+    { WireFormat::mls_plaintext, content, context },
+    auth,
+  });
+
+  return suite.digest().hmac(membership_key, tbm);
+}
+
+tls::ostream&
+operator<<(tls::ostream& str, const MLSPlaintext& obj)
+{
+  str << obj.content << obj.auth;
+
+  if (obj.content.sender.sender_type() == SenderType::member) {
+    str << opt::get(obj.membership_tag);
+  }
+
+  return str;
+}
+
+tls::istream&
+operator>>(tls::istream& str, MLSPlaintext& obj)
+{
+  str >> obj.content;
+
+  obj.auth.content_type = obj.content.content_type();
+  str >> obj.auth;
+
+  if (obj.content.sender.sender_type() == SenderType::member) {
+    obj.membership_tag.emplace();
+    str >> opt::get(obj.membership_tag);
```

Comment:
Refactor as `switch (obj.content.sender_type())`

Reply:
Actually, this doesn't work well, since the logic to set `auth.content_type` would have to get replicated.